internal/graph/lookupsubjects.go

```diff
@@ -495,114 +905,389 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
+		return "", nil
+	}
+
+	return sortedSubjectIDs[len(sortedSubjectIDs)-1], nil
 }
 
-func (lse *lookupSubjectsExclusion) CompletedChildOperations() error {
-	var foundSubjects datasets.SubjectSetByResourceID
-	metadata := emptyMetadata
+// createFilteredAndLimitedResponse creates a filtered and limited (as is necessary via the cursor and limits)
+// version of the subjects, returning a DispatchLookupSubjectsResponse ready for publishing with just that
+// subset of results.
+func createFilteredAndLimitedResponse(
+	ci cursorInformation,
+	subjects map[string]*v1.FoundSubjects,
+	metadata *v1.ResponseMeta,
+) (*v1.DispatchLookupSubjectsResponse, func(), error) {
+	if subjects == nil {
+		return nil, func() {}, spiceerrors.MustBugf("nil subjects given to createFilteredAndLimitedResponse")
+	}
 
-	for index := 0; index < len(lse.collectors); index++ {
-		collector := lse.collectors[index]
-		results := datasets.NewSubjectSetByResourceID()
-		for _, result := range collector.Results() {
-			metadata = combineResponseMetadata(metadata, result.Metadata)
-			if err := results.UnionWith(result.FoundSubjectsByResourceId); err != nil {
-				return fmt.Errorf("failed to UnionWith under lookupSubjectsExclusion: %w", err)
+	afterSubjectID, _ := ci.headSectionValue()
+
+	// Filter down the subjects found by the cursor (if applicable) and the apply a limit.
+	filteredSubjectIDs := mapz.NewSet[string]()
+	for _, foundSubjects := range subjects {
+		for _, foundSubject := range foundSubjects.FoundSubjects {
+			// NOTE: wildcard is always returned, because it is needed by all branches, at all times.
+			if foundSubject.SubjectId == tuple.PublicWildcard || (afterSubjectID == "" || foundSubject.SubjectId > afterSubjectID) {
+				filteredSubjectIDs.Add(foundSubject.SubjectId)
 			}
 		}
+	}
 
-		if index == 0 {
-			foundSubjects = results
-		} else {
-			foundSubjects.SubtractAll(results)
-			if foundSubjects.IsEmpty() {
-				return nil
-			}
+	sortedSubjectIDs := filteredSubjectIDs.AsSlice()
+	sort.Strings(sortedSubjectIDs)
+
+	subjectIDsToPublish := make([]string, 0, len(sortedSubjectIDs))
+	subjectIDsToPublishWithoutWildcard := make([]string, 0, len(sortedSubjectIDs))
+
+	done := func() {}
+	for _, subjectID := range sortedSubjectIDs {
+		// Wildcards are always published, regardless of the limit.
+		if subjectID == tuple.PublicWildcard {
+			subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+			continue
 		}
+
+		ok := ci.limits.prepareForPublishing()
+		if !ok {
+			break
+		}
+
+		subjectIDsToPublish = append(subjectIDsToPublish, subjectID)
+		subjectIDsToPublishWithoutWildcard = append(subjectIDsToPublishWithoutWildcard, subjectID)
 	}
 
-	return lse.parentStream.Publish(&v1.DispatchLookupSubjectsResponse{
-		FoundSubjectsByResourceId: foundSubjects.AsMap(),
+	if len(subjectIDsToPublish) == 0 {
+		return nil, done, nil
+	}
+
+	// Determine the subject ID for the cursor. If there are any concrete subject IDs, then the last
```

Comment:
We always sort the returned results so the wildcard is last when retrieving from the datastore. This occurs on line 1193 after we union together the wildcard with the concrete results found. Thus, since we sort at every level (and wildcard always sorts last), we're guaranteed that it should always appear at the end. If you like I can add additional point unit tests for the individual reducers to validate this, but we can't really test everything directly

Reply:
after reverse engineering the code from scratch instead of looking at the GitHub diff, I got a better understanding, thanks!